File: NodeJS/translateRNA.js
//import modules
const { parse } = require('csv-parse') //parsing csv
const fs = require('fs') //read from file system
const ps = require('prompt-sync') //get user input

//path
const file = "C:\\Users\\rorop\\Desktop\\Script_In_All_Languages\\codon_table.csv"

//read from csv and append to list, as a list of lists
function createReadStream() {
    var csvData = [];
    return new Promise(resolve => {
        fs.createReadStream(file)
            //delimiter = ',' and ignore header by starting at line 2
            .pipe(parse({delimiter: ',', from_line: 2}))
            .on('data', function (csvrow) {
                // console.log(csvrow);
                csvData.push(csvrow);
            })
            .on('end', () => resolve(csvData));
    });
}

//create dictionary from list of list
function create_dict(list_of_lists, key_index, value_index){
    var dict = {}
    for (i = 0; i < list_of_lists.length; i++){
        dict[list_of_lists[i][key_index]] = list_of_lists[i][value_index]
    }
    return dict
}

//clean dict; replace "Stop" by "_"
function clean_dict(dict){
    //TODO understand this script
    Object.keys(dict)
        .forEach(function(key) {
            if (dict[key] == 'Stop') {
                dict[key] = '_'
            }
        }
    )
    return dict
}

//get user input
function get_user_seq(){
    const prompt = ps()
    let seq = prompt('Enter RNA sequence: ')

    seq = seq.toUpperCase()

    var accepted = ['A', 'U', 'C', 'G']

    for (const character of seq) {
        if (!accepted.includes(character)) {
            throw new Error('Only (A,U,C,G) are accepted');
        }
    }
    return seq
}

//slice into codons
function chunkString (str, len) {
    //TODO understand this script
    const size = Math.ceil(str.length/len)
    const r = Array(size)
    let offset = 0
    for (let i = 0; i < size; i++) {
        r[i] = str.substr(offset, len)
        offset += len
    }
    return r
}

//translate
function translate(seq, dict){
    var protein = ""

    for (i=0; i < seq.length; i++){
        protein += dict[seq[i]]
    }
    return protein
}


async function main() {
    const csvData = await createReadStream();

    // LOGIC AFTER GETTING DATA
    var amino_dict = create_dict(csvData, 0, 2)
    var amino_clean_dict = clean_dict(amino_dict)
    var seq = get_user_seq()
    var seq_chunked = chunkString(seq, 3)
    console.log(translate(seq_chunked, amino_clean_dict))

}

main()

File: translateRNA.py
#Read file and return list
def read_file(file):

    #Open file as read only and read every line while removing '\n'
    with open(file, 'r') as f:
        rows = f.read().splitlines()

    #all_rows as a list of list
    all_rows = []
    #Loop through rows (starting from 2nd, to ignore header)
    for row in rows[1:]:
        #Split row based on comma
        all_rows.append(row.split(','))

    return all_rows

#Create dictionary
def create_dict(rows):

    #Store 1st item of the row as the key (codon)
    #Store 3rd item of the row as the value (amino acid)
    amino_codon_dict = {}
    for row in rows:
        codon = row[0]
        amino_acid_letter = row[2]

        #Transform "Stop" into "_"
        if amino_acid_letter == "Stop":
            amino_acid_letter = "_"

        amino_codon_dict[codon] = amino_acid_letter

    return amino_codon_dict

def turnUppercase(seq):
    return seq.upper()

#Check if user entered a valid RNA sequence (no letters other than A,U,G,C)
def validate_sequence(seq):

    accepted = ['A', 'U', 'C', 'G']

    for letter in seq:
        if letter not in accepted:
            raise Exception("Sequence contains an faulty letter (one that is not A, U, C, G)")

#Splits string into list of substrings separated by n characters
def split_into_nth(a_string, nth):

    #Store the substring
    the_list = []
    for i in range(0, len(a_string), nth):
        substring = a_string[i:i+nth]
        the_list.append(substring)
    return the_list

#Remove codons that are incomplete (less than 3 nucleotides)
def remove_incomplete_codons(list_of_codons):
    for codon in list_of_codons:
        if len(codon) < 3:
            list_of_codons.remove(codon)

    return list_of_codons

#Translate given sequence with previously created dictionary
def translate_RNA_to_Protein(list_of_codons):

    protein = ""
    for codon in list_of_codons:
        try:
            protein += amino_dict[codon]
        except KeyError:
            print("Codon not found in the table")
    return protein

#Read file and rows
file = r'C:\Users\rorop\Desktop\Script_In_All_Languages\codon_table.csv'
file_read = read_file(file)

#Create dictionary
amino_dict = create_dict(file_read)

#Ask user for RNA and turn to uppercase
user_sequence = input("Enter your DNA sequence: ")
user_sequence_upper = turnUppercase(user_sequence)

#Validate sequence
validate_sequence(user_sequence_upper)

#Split sequence into codons and remove codons that aren't 3 letters
seq_split = split_into_nth(user_sequence_upper, 3)
seq_complete = remove_incomplete_codons(seq_split)

#Translate sequence into protein
print(translate_RNA_to_Protein(seq_complete))

